### PR TITLE
support pytest `python_functions` config via a pytest plugin

### LINKED ISSUE #155: request to support pytest configured with other naming conventions other than "test_"
Pytest has a configuration option to change the default so that instead of tests being named "test_" you can change it to "check_" or "should_" or similar. If you do that, approvals stops working. If pytest can find and execute your test I would expect approvals to also still work and evaluate whether the test passed or not.

This is the docs for pytest:

https://docs.pytest.org/en/stable/example/pythoncollection.html#change-naming-conventions

With this pytest.ini:

[pytest]
python_functions = *_check

and this test code:

def instance_check():
    result = str(Hiker().instance_answer())
    verify(result)
 
then you get an error like this:

Could not find test method/function. Possible reasons could be:
    1) approvaltests is not being used inside a test function
    2) your test framework is not supported by ApprovalTests (unittest and pytest are currently supported)."""
>           raise FrameNotFound(message)
E           approvaltests.approval_exception.FrameNotFound: Could not find test method/function. Possible reasons could be: 
E           1) approvaltests is not being used inside a test function 
E           2) your test framework is not supported by ApprovalTests (unittest and pytest are currently supported).

/usr/local/lib/python3.10/site-packages/approvaltests/core/namer.py:77: FrameNotFound

I have this cyber-dojo instance set up with this problem:

https://cyber-dojo.org/kata/edit/1cMGrz

### PR DESCRIPTION
**Pull request by nitsanavni**
fixes #155 "on my machine"

### TODO
- [ ] add tests

### CODE REVIEW

**Comment by nitsanavni** [comment]
@emilybache please review

File: approvaltests/namer/stack_frame_namer.py [hunk 1]
```diff
@@ -51,10 +71,15 @@ def is_test_method(frame: FrameInfo) -> bool:
...
 
-        is_pytest_test = method_name.startswith("test_")
+    @staticmethod
+    def is_test_method(frame: FrameInfo) -> bool:
+        method_name = frame[3]
 
-        return is_unittest_test or is_pytest_test
+        return StackFrameNamer.is_unittest_test(
+            frame
+        ) or StackFrameNamer.is_pytest_test(method_name)
```
Comment (emilybache) [inline approvaltests/namer/stack_frame_namer.py:82]: it's bothering me that the StackFrameNamer.is_unittest_test and is_pytest_test take different arguments. I'd pass the frame to both.

File: approvaltests/namer/stack_frame_namer.py [hunk 2]
```diff
@@ -40,7 +42,25 @@ def get_test_frame_index(caller: List[FrameInfo]) -> int:
...
+        patterns = PytestConfig.test_naming_patterns
+        
+        # taken from pytest/python.py (class PyCollector)
+        for pattern in patterns:
+            if method_name.startswith(pattern):
+                return True
+                # Check that name looks like a glob-string before calling fnmatch
+                # because this is called for every name in each collected module,
+                # and fnmatch is somewhat expensive to call.
+            elif ("*" in pattern or "?" in pattern or "[" in pattern) and fnmatch.fnmatch(
```
Comment (emilybache) [inline approvaltests/namer/stack_frame_namer.py:55]: Is this the same code that pytest uses internally? Is there any way to call the same code in pytest as a library?

File: approvaltests/pytest/pytest_plugin.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+from .pytest_config import set_pytest_config
+
+def pytest_configure(config):
```
Comment (emilybache) [inline approvaltests/pytest/pytest_plugin.py:3]: I wasnt sure how this was all wired together - what calls pytest_configure?

**Comment by emilybache** [comment]
I'm a little concerned that this code might solve the immediate problem but not in a sustainable way. It seems to be duplicating functionality in pytest, by reading pytest's configuration file. I wonder if there is a way to re-use the configuration file parsing code from pytest itself?

**Comment by nitsanavni** [comment]
@emilybache Thanks for reviewing!

> duplicating functionality in pytest

Indeed it does, but does not rely on the implementation but rather on the public pytest configuration interface.
I've considered calling pytest internals, but that would make our code depend on implementation details rather than on public interfaces which tend to be more stable. (For example I've noticed that the specific required code has moved between different files and classes between different pytest versions)

Copying the code was an easy way to avoid rewriting it, but since there's no obvious way to DRY it I was ok with keeping it as is. I do think that this code merits some unit testing on our side, though.

The parsing of the `pytest.ini` file is still done by pytest, not us. We merely use the plugin mechanism (which is also a public interface of pytest) to access the specific config field of interest.

File: approvaltests/namer/stack_frame_namer.py (hunk 2 above)
Comment (nitsanavni) [reply approvaltests/namer/stack_frame_namer.py:55]: Indeed, good eye.
See discussion [here](https://github.com/approvals/ApprovalTests.Python/pull/157#issuecomment-1884369703).

**Comment by emilybache** [comment]
That is a good response to my concern. It's good to be relying on a public pytest interface. The amount of duplicated code is small and hopefully relatively stable since it refers to the wildcards used in teh published format of the pytest.ini file.

File: approvaltests/pytest/pytest_plugin.py (hunk 3 above)
Comment (nitsanavni) [reply approvaltests/pytest/pytest_plugin.py:3]: This is a pytest plugin, I think pytest treats any function starting with `pytest_` in a plugin module as an entry point to the plugin and sends the config as a fixture.

**Comment by emilybache** [comment]
looks good to me!

**Comment by JulieGunawan** [comment]
Potential starting point for writing test:
https://github.com/nitsanavni/ApprovalTests.Python/tree/test-example

**Pull request merged**